Add remote datasource

n-h-diaz opened the pull request:
Right now I only added it when "useSpannerGraph" is also true, to avoid affecting other instances while still in development.

I wasn't sure the best way to test this. One options is to create the TypeMock DataSource and use that in place of TypeRemote in tests, but this wouldn't really test the remote mixer (but instead could help test merging behavior). Let me know if you have any ideas or suggestions! 

This PR also adds the wrapper for V3 resolve, but this currently doesn't return anything.

keyurva commented on cmd/main.go:
```diff
@@ -265,6 +266,17 @@ func main() {
 
+	// Remote Mixer.
+	// TODO: remove flag for Spanner Graph after validating and adding merging.
+	if *useSpannerGraph && *remoteMixerDomain != "" {
```
What do you think about an `enableV3` flag to test with different combinations of data sources?

n-h-diaz replied:
Added this flag for each datasource

keyurva commented on internal/server/remote/client.go:
```diff
@@ -0,0 +1,75 @@
+	"github.com/datacommonsorg/mixer/internal/util"
+)
+
+// RemoteClient encapsulates the Remote client.
```
nit: consider making a mention of a remote mixer for clarity.

n-h-diaz replied:
Done

keyurva commented on internal/server/remote/client.go:
```diff
@@ -0,0 +1,75 @@
+// NewRemoteClient creates a new RemoteClient.
+func NewRemoteClient(metadata *resource.Metadata, httpClient *http.Client) (*RemoteClient, error) {
+	if metadata.RemoteMixerDomain == "" || metadata.RemoteMixerAPIKey == "" {
+		return nil, fmt.Errorf("error creating remote client")
```
Consider adding a reason that can help users fix it - something about remote url and key must be specified.

n-h-diaz replied:
Done

keyurva commented on internal/server/remote/client.go:
```diff
@@ -0,0 +1,75 @@
+
+func (rc *RemoteClient) Node(req *pbv3.NodeRequest) (*pbv3.NodeResponse, error) {
+	resp := &pbv3.NodeResponse{}
+	err := util.FetchRemote(rc.metadata, rc.httpClient, "/v3/node", req, resp)
```
Wondering if these (except the new end points) should be to `/v2`?

n-h-diaz replied:
I switched these to V2 (so had to add some conversion between V2 and V3 protos). If we do this though, we should just be careful that V3 response aligns with expected V2 response so they can be merged

keyurva reviewed (commented):
This is very well done!

> I wasn't sure the best way to test this

Could using the `enableV3` approach I mentioned in a comment help with this?

n-h-diaz commented:
> This is very well done!
> 
> > I wasn't sure the best way to test this
> 
> Could using the `enableV3` approach I mentioned in a comment help with this?

I added the flag and enabled in autopush. For testing, it looks like remote mixer currently is tested either locally or by mocking results (since it requires an API key). I added a test for merging node and can continue to add more as we add more merging functionality. Let me know if you have any suggestions for how I should better test this though!

keyurva commented on internal/server/remote/client.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+func (rc *RemoteClient) Node(req *pbv3.NodeRequest) (*pbv3.NodeResponse, error) {
+	v2Req := &pbv2.NodeRequest{
```
nit and optional: Consider using helper methods for doing conversions.

n-h-diaz replied:
Good idea, done

keyurva commented:
> > This is very well done!
> > > I wasn't sure the best way to test this
> > 
> > 
> > Could using the `enableV3` approach I mentioned in a comment help with this?
> 
> I added the flag and enabled in autopush. For testing, it looks like remote mixer currently is tested either locally or by mocking results (since it requires an API key). I added a test for merging node and can continue to add more as we add more merging functionality. Let me know if you have any suggestions for how I should better test this though!

Hmm. Wondering if a cloudbuild that runs at a regular cadence is something to consider here. We can expand the scope as an overall V3 sanity test. 

Adding @hqpho - if they have any suggestions.

[All of this is of course outside the scope of this PR and we can take the discussion offline if needed.]
